File: src-tauri/src/commands.rs
use tauri::{AppHandle, Manager, State, WebviewUrl, WebviewWindowBuilder};

use crate::fs;
use crate::types::{DirEntry, FileRead, FileWritten, FsError};
use crate::AppState;

#[tauri::command]
pub fn fs_read(path: String) -> Result<FileRead, FsError> {
    fs::read(&path)
}

#[tauri::command]
pub fn fs_write(path: String, content: String) -> Result<FileWritten, FsError> {
    fs::write(&path, &content)
}

#[tauri::command]
pub fn fs_create(path: String) -> Result<(), FsError> {
    fs::create(&path)
}

#[tauri::command]
pub fn fs_rename(from: String, to: String) -> Result<(), FsError> {
    fs::rename(&from, &to)
}

#[tauri::command]
pub fn fs_copy(from: String, to: String) -> Result<(), FsError> {
    fs::copy(&from, &to)
}

#[tauri::command]
pub fn fs_list(path: String) -> Result<Vec<DirEntry>, FsError> {
    fs::list(&path)
}

#[tauri::command]
pub fn fs_delete(path: String) -> Result<(), FsError> {
    fs::delete(&path)
}

#[tauri::command]
pub fn fs_count_recursive(path: String) -> Result<u64, FsError> {
    fs::count_recursive(&path)
}

/// Reveal a file or folder in Finder (highlights it inside its parent
/// directory). macOS-only. Equivalent to right-click → Show in Finder.
#[tauri::command]
pub fn shell_reveal_in_finder(path: String) -> Result<(), FsError> {
    std::process::Command::new("open")
        .args(["-R", &path])
        .spawn()
        .map(|_| ())
        .map_err(|e| FsError::Io(format!("reveal failed: {e}")))
}

/// Open Terminal.app at the given directory (or the parent directory if
/// `path` points at a file). macOS-only.
#[tauri::command]
pub fn shell_open_in_terminal(path: String) -> Result<(), FsError> {
    let p = std::path::Path::new(&path);
    let dir = if p.is_dir() {
        p.to_path_buf()
    } else {
        p.parent()
            .map(|x| x.to_path_buf())
            .unwrap_or_else(|| p.to_path_buf())
    };
    std::process::Command::new("open")
        .args(["-a", "Terminal", &dir.to_string_lossy()])
        .spawn()
        .map(|_| ())
        .map_err(|e| FsError::Io(format!("open terminal failed: {e}")))
}

#[tauri::command]
pub fn watcher_subscribe(path: String, state: State<AppState>) -> Result<(), FsError> {
    state
        .watcher
        .lock()
        .map_err(|_| FsError::Io("watcher mutex poisoned".into()))?
        .subscribe(&path)
        .map_err(|e| FsError::Io(e.to_string()))
}

#[tauri::command]
pub fn window_open_preview(
    app: AppHandle,
    label: String,
    title: String,
    doc_id: String,
) -> Result<(), FsError> {
    if app.get_webview_window(&label).is_some() {
        return Ok(());
    }
    let url = WebviewUrl::App(format!("preview.html?docId={}", doc_id).into());
    WebviewWindowBuilder::new(&app, label, url)
        .title(title)
        .inner_size(900.0, 700.0)
        .build()
        .map(|_| ())
        .map_err(|e| FsError::Io(e.to_string()))
}

#[tauri::command]
pub fn window_close(app: AppHandle, label: String) -> Result<(), FsError> {
    if let Some(w) = app.get_webview_window(&label) {
        w.close().map_err(|e| FsError::Io(e.to_string()))?;
    }
    Ok(())
}

/// Exit the whole process cleanly. Used by future menu items like Cmd-Q;
/// not on the red-X path (native close handles that directly now that the
/// window-event intercept is gone).
#[tauri::command]
pub fn app_exit(app: AppHandle) {
    app.exit(0);
}

/// Content of the bundled welcome / test document. Compiled into the binary
/// so the installed .app is self-contained — no external resource lookup,
/// no path-relative fragility.
const WELCOME_CONTENT: &str = include_str!("../resources/welcome.md");

/// On first run, seeds a "Welcome.md" file in the user's Documents folder
/// and returns its path. Subsequent calls return the same path without
/// overwriting, so the user's edits stick.
#[tauri::command]
pub fn ensure_welcome_file(app: AppHandle) -> Result<String, FsError> {
    let docs = app
        .path()
        .document_dir()
        .map_err(|e| FsError::Io(format!("document_dir: {}", e)))?;
    let app_dir = docs.join("Yeogi .MD Editor");
    std::fs::create_dir_all(&app_dir).map_err(|e| FsError::Io(e.to_string()))?;
    let file_path = app_dir.join("Welcome.md");
    if !file_path.exists() {
        std::fs::write(&file_path, WELCOME_CONTENT).map_err(|e| FsError::Io(e.to_string()))?;
    }
    Ok(file_path.to_string_lossy().to_string())
}

/// Overwrite "Welcome.md" with the bundled seed. Unlike `ensure_welcome_file`,
/// this unconditionally replaces the file — used by the "Help → Reset
/// Welcome.md to Default" menu item so users can pick up the latest seed
/// after an app update. Callers are expected to have already shown the user
/// a destructive-confirmation prompt before invoking this.
#[tauri::command]
pub fn reseed_welcome_file(app: AppHandle) -> Result<String, FsError> {
    let docs = app
        .path()
        .document_dir()
        .map_err(|e| FsError::Io(format!("document_dir: {}", e)))?;
    let app_dir = docs.join("Yeogi .MD Editor");
    std::fs::create_dir_all(&app_dir).map_err(|e| FsError::Io(e.to_string()))?;
    let file_path = app_dir.join("Welcome.md");
    std::fs::write(&file_path, WELCOME_CONTENT).map_err(|e| FsError::Io(e.to_string()))?;
    Ok(file_path.to_string_lossy().to_string())
}

/// Rebuild the native menu to reflect the frontend's current state and swap
/// it in. Called once at mount (after preferences hydrate) and any time
/// either the MRU list or the theme preference changes. Batching both into
/// one command avoids the Rust side needing to cache either piece of state
/// — the frontend already does, and sends both on every change.
#[tauri::command]
pub fn sync_menu_state(
    app: AppHandle,
    recent_files: Vec<String>,
    theme: String,
) -> Result<(), FsError> {
    let entries: Vec<crate::menu::RecentFile> = recent_files
        .into_iter()
        .map(|path| {
            let display = std::path::Path::new(&path)
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or(&path)
                .to_string();
            crate::menu::RecentFile { path, display }
        })
        .collect();
    let menu = crate::menu::build_menu(&app, &entries, theme.as_str())
        .map_err(|e| FsError::Io(format!("menu build failed: {}", e)))?;
    app.set_menu(menu)
        .map_err(|e| FsError::Io(format!("set_menu failed: {}", e)))?;
    Ok(())
}

File: src/lib/ipc/commands.ts
import { invoke } from "@tauri-apps/api/core";
import type { DirEntry, FsError } from "./types";

// The generated ./types has mtime_ms: bigint because ts-rs 8 maps Rust i64
// that way. At runtime the wire value is always a plain JSON number, so the
// boundary re-types it as number. See PLAN.md Phase 3 preamble.
export interface FileRead {
  content: string;
  mtime_ms: number;
  encoding: string;
}

export interface FileWritten {
  mtime_ms: number;
}

export async function fsRead(path: string): Promise<FileRead> {
  return (await invoke("fs_read", { path })) as unknown as FileRead;
}

export async function fsWrite(path: string, content: string): Promise<FileWritten> {
  return (await invoke("fs_write", { path, content })) as unknown as FileWritten;
}

export async function fsCreate(path: string): Promise<void> {
  return invoke("fs_create", { path });
}

export async function fsRename(from: string, to: string): Promise<void> {
  return invoke("fs_rename", { from, to });
}

/**
 * File-only copy. Refuses to overwrite an existing target — callers must
 * compute a unique destination path. Used by the file-tree Duplicate
 * action.
 */
export async function fsCopy(from: string, to: string): Promise<void> {
  return invoke("fs_copy", { from, to });
}

export async function fsList(path: string): Promise<DirEntry[]> {
  return invoke("fs_list", { path });
}

/**
 * Permanently remove a file or folder. For folders, recursively removes
 * every descendant. The caller MUST gate this behind a destructive-
 * confirmation modal — the Rust command does no extra prompting.
 */
export async function fsDelete(path: string): Promise<void> {
  return invoke("fs_delete", { path });
}

/**
 * Count every descendant of `path` (files + directories combined),
 * excluding `path` itself. Returns 0 for a file. Used by the delete
 * confirmation dialog to surface item-count.
 */
export async function fsCountRecursive(path: string): Promise<number> {
  return invoke("fs_count_recursive", { path });
}

/** Reveal a file/folder in Finder (highlights inside its parent dir). */
export async function shellRevealInFinder(path: string): Promise<void> {
  return invoke("shell_reveal_in_finder", { path });
}

/**
 * Open Terminal.app at `path`. If `path` is a file, opens at the file's
 * parent directory.
 */
export async function shellOpenInTerminal(path: string): Promise<void> {
  return invoke("shell_open_in_terminal", { path });
}

export async function watcherSubscribe(path: string): Promise<void> {
  return invoke("watcher_subscribe", { path });
}

/**
 * First-run welcome file seeding. On the Rust side, copies the bundled
 * welcome.md into ~/Documents/Yeogi .MD Editor/Welcome.md (idempotent —
 * never overwrites) and returns its absolute path.
 */
export async function ensureWelcomeFile(): Promise<string> {
  return invoke("ensure_welcome_file");
}

/**
 * Overwrite ~/Documents/Yeogi .MD Editor/Welcome.md with the bundled seed,
 * replacing whatever's there. Returns the file's path. Callers must
 * prompt the user for destructive confirmation before invoking — the Rust
 * command itself trusts its caller.
 */
export async function reseedWelcomeFile(): Promise<string> {
  return invoke("reseed_welcome_file");
}

/**
 * Rebuild the native menu to reflect the current `recentFiles` list and
 * `theme` preference. Called at mount and on every change to either. One
 * command (rather than two) means either change triggers the same server-
 * side rebuild path, keeping the menu in a known-consistent state.
 */
export async function syncMenuState(
  recentFiles: string[],
  theme: string,
): Promise<void> {
  return invoke("sync_menu_state", { recentFiles, theme });
}

export type { DirEntry, FsError };
